pandas/core/groupby/generic.py

```diff
@@ -1710,3 +1703,34 @@ def _normalize_keyword_aggregation(kwargs):
         order.append((column,
                       com.get_callable_name(aggfunc) or aggfunc))
     return aggspec, columns, order
+
+
+def _recast_datetimelike_result(result: DataFrame) -> DataFrame:
+    """
+    If we have date/time like in the original, then coerce dates
+    as we are stacking can easily have object dtypes here.
+
+    Parameters
+    ----------
+    result : DataFrame
+
+    Returns
+    -------
+    DataFrame
+
+    Notes
+    -----
+    - Assumes Groupby._selected_obj has ndim==2 and at least one
+    datetimelike column
+    """
+    result = result.copy()
+
+    ocols = [idx for idx in range(len(result.columns))
+             if is_object_dtype(result.dtypes[idx])]
+
+    for cidx in ocols:
+        cvals = result.iloc[:, cidx].values
+        result.iloc[:, cidx] = maybe_convert_objects(cvals,
```

Comment:
this might be quite a performance hit as before we were doing a single assignment, here each column is done.

can you restructure this to something like

```
result.iloc[:, ocols] = [maybe_convert_object(result.iloc[:, cidx].values, convert_numeric=False) for cidx in ocols]
```

ideally better name for ocols, maybe object_columns_indexer

Reply:
It can't really be a performance hit since it raises in master.  Better name: will do.  single assignment... I find less clear, but OK.